File: lib/supplier/interfaces/SupplierOrderClient.ts
/**
 * Supplier Order Client Interface
 * 
 * This interface defines the contract for order management with suppliers.
 * Following SOLID principles, this allows us to add new suppliers without
 * modifying existing core business logic.
 * 
 * Interface Segregation Principle: Only includes order-related operations
 * Dependency Inversion Principle: High-level modules depend on this abstraction
 */

export interface OrderItem {
  supplierSku: string;
  quantity: number;
  unitPrice: number;
}

export interface ShippingAddress {
  name: string;
  email: string;
  phone?: string;
  address1: string;
  address2?: string;
  city: string;
  state: string;
  postalCode: string;
  country: string;
}

export interface DropshipOrderRequest {
  items: OrderItem[];
  shippingAddress: ShippingAddress;
  customerNotes?: string;
}

export interface DropshipOrderResponse {
  success: boolean;
  supplierOrderId?: string;
  trackingNumber?: string;
  estimatedDelivery?: Date;
  totalAmount: number;
  shippingCost: number;
  errors: string[];
}

export interface OrderStatusResponse {
  supplierOrderId: string;
  status: string;
  trackingNumber?: string;
  shippedAt?: Date;
  deliveredAt?: Date;
  notes?: string;
}

export interface SupplierOrderClient {
  /**
   * Place a dropship order with the supplier
   */
  placeOrder(orderRequest: DropshipOrderRequest): Promise<DropshipOrderResponse>;
  
  /**
   * Get order status from supplier
   */
  getOrderStatus(supplierOrderId: string): Promise<OrderStatusResponse>;
  
  /**
   * Cancel an order (if supported by supplier)
   */
  cancelOrder(supplierOrderId: string): Promise<boolean>;
  
  /**
   * Get tracking information
   */
  getTrackingInfo(supplierOrderId: string): Promise<string | null>;
  
  /**
   * Test order placement (sandbox mode)
   */
  testOrder(): Promise<boolean>;
}

File: ds3-dropshipping-service/src/main/java/com/ds3/dropshipping/supplier/interfaces/SupplierOrderClient.java
package com.ds3.dropshipping.supplier.interfaces;

import com.ds3.dropshipping.supplier.model.DropshipOrderResponse;
import com.ds3.dropshipping.supplier.model.DropshipOrderRequest;
import com.ds3.dropshipping.supplier.model.OrderStatusResponse;

import java.util.concurrent.CompletableFuture;

/**
 * Supplier Order Client Interface
 * 
 * This interface defines the contract for order management with suppliers.
 * Following SOLID principles, this allows us to add new suppliers without
 * modifying existing core business logic.
 * 
 * Interface Segregation Principle: Only includes order-related operations
 * Dependency Inversion Principle: High-level modules depend on this abstraction
 */
public interface SupplierOrderClient {

    /**
     * Place a dropship order with the supplier
     * 
     * @param orderRequest The order request details
     * @return CompletableFuture containing order response
     */
    CompletableFuture<DropshipOrderResponse> placeOrder(DropshipOrderRequest orderRequest);

    /**
     * Get order status from supplier
     * 
     * @param supplierOrderId The supplier's order ID
     * @return CompletableFuture containing order status
     */
    CompletableFuture<OrderStatusResponse> getOrderStatus(String supplierOrderId);

    /**
     * Cancel an order (if supported by supplier)
     * 
     * @param supplierOrderId The supplier's order ID
     * @return CompletableFuture containing true if cancellation successful
     */
    CompletableFuture<Boolean> cancelOrder(String supplierOrderId);

    /**
     * Get tracking information
     * 
     * @param supplierOrderId The supplier's order ID
     * @return CompletableFuture containing tracking number or null
     */
    CompletableFuture<String> getTrackingInfo(String supplierOrderId);

    /**
     * Test order placement (sandbox mode)
     * 
     * @return CompletableFuture containing true if test order successful
     */
    CompletableFuture<Boolean> testOrder();
}
